messaging/models.py

```diff
@@ -176,8 +177,27 @@ def get_recipient_guardians(self):
                 events__occurrences__enrolments__child=OuterRef("pk")
             )
 
-            children_with_invitation = children.filter(
-                Q(Exists(enrollable_single_events) | Exists(enrollable_event_groups))
+            enrolled_this_year = Coalesce(
+                Subquery(
+                    enrolments.filter(
+                        occurrence__time__year=now.year, child=OuterRef("pk")
+                    )
+                    .values("child")
+                    .annotate(count=Count("pk"))
+                    .values("count")
```

Comment:
Okay, after rerunning the whole test set, I found out that the `.values("child")` really is needed 😄 It relates to `child=OuterRef("pk")` doesn't it?

Reply:
Indeed `count()` cannot be used since that's something that Django tries to resolve immediately. We want that to be resolved in the DB, since we use the count for filtering.

> So wouldn't the coalesce-function always select the first value, because it's always a non-null value? https://docs.djangoproject.com/en/4.0/ref/models/database-functions/#coalesce

Maybe it would be more clear if the coalesce was in the main query. The problem is that sometimes it is a `NULL` value. Children which do not have any enrollments will get a `NULL` value when annotating the value `enrolled_this_year`. This coalesce turns that `NULL` into `0`. Reasoning for coalesce is that the `enrolled_this_year__gte` doesn't work with null values. 

As an option for the `Coalesce` we can e.g. do:
```python
.exclude(enrolled_this_year__isnull=False, enrolled_this_year__gte=F("project__enrolment_limit"))
```
but imho coalsce version makes it more understandable. I have not tested if the coalesce makes the query less performant, but given that it only adds a default value, I would assume it's unnoticeable.

The `.values("child")` is indeed required, it groups the query result. Some details are found [here](https://stackoverflow.com/a/43771738).